SOS/dotnet-dump extensibility - Part 2

mikem8361 opened the pull request:
Native SOS clean up and refactoring

Added and updated IHost, ITarget and IRuntime native interfaces and
implementations. Going forward these will be provided by the managed
host support.

Added "runtimes" command and moved the runtime selection options from
sosstatus to this command.

Move include files to common SOS inc directory.

hoyosjs commented on src/SOS/Strike/util.h:
```diff
@@ -38,8 +33,11 @@ inline void RestoreSOToleranceState() {}
 #include "cordebug.h"
 #include "static_assert.h"
 #include <string>
+#include <releaseholder.h>
```
Can we use the quote pattern for files we own please?

mikem8361 replied:
Unless I'm mistaken, quotes means local files in the same directory as the source and <h> means to use the search directory paths.  The meaning has probably change over the years but I have always thought of it as that.

mikem8361 replied:
Done

hoyosjs commented on src/SOS/Strike/strike.cpp:
```diff
@@ -16189,31 +16137,19 @@ HRESULT SetNGENCompilerFlags(DWORD flags)
-    }
-    else if (FAILED(pCorDebugProcess->QueryInterface(__uuidof(ICorDebugProcess2), (void**) &proc2)))
+    // if CLR is already loaded, try to change the flags now
+    hr = GetTarget()->GetRuntime(&g_pRuntime);
```
hmm. Shouldn't initialization be done in a common path instead of initializing the runtime every time?

mikem8361 replied:
This is a special case because SetNGENCompilerFlags gets called from the EventCallbacks::LoadModule and it is historical. Normally for commands that use the INIT_API_EXT macro will get g_pRuntime setup.  

This eventually could be cleaned up but I don't want to do it in this PR.

hoyosjs commented on src/SOS/Strike/ExpressionNode.cpp:
```diff
@@ -50,10 +50,10 @@ HRESULT ExpressionNode::CreateExpressionNode(__in_z WCHAR* pExpression, Expressi
     HRESULT Status = g_pRuntime->GetCorDebugInterface(&s_pCorDebugProcess);
-    if (FAILED(Status)) 
-    {
+    if (FAILED(Status)) {
```
Nit style for the brace

mikem8361 replied:
I know it isn't exactly our convention but when there is one line in the braces I break the rules instead of not putting braces at all (I really hate that).

hoyosjs commented:
>             0x9d, 0xe6, 0x57, 0x34, 0xc3, 0xfe, 0xd6, 0x89);

Do we need to check this one in? Shouldn't this just be pulled from an SDK in the machine?

---
Refers to: src/SOS/inc/dbgeng/dbgeng.h:49 in cb973fd. [](commit_id = cb973fdea519ca0d8a8ea7c366d3bcdbffef97b5, deletion_comment = False)

mikem8361 commented:
They have always been checked in.  I just moved them and updated them.  Can we always assume that there will be a new enough SDK installed on our lab machines (and everybody's dev machines)?  How do you reference this SDK from cmake?

hoyosjs commented on src/SOS/Strike/platform/targetimpl.cpp:
```diff
@@ -0,0 +1,350 @@
+\**********************************************************************/
+void Target::DisplayStatusInstance()
+{
+    static const char* osName[] = {
```
We should probably move this next to the `ITarget::OperatingSystem` definition to ensure they always match.

mikem8361 replied:
The target.h/ITarget interface is the abstraction. I don't think implementation details like this belong there.

hoyosjs commented:
For the SDK, there's this. But given the change is big, probably we can clean that up later: https://cmake.org/cmake/help/v3.9/variable/CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION.html

mikem8361 commented on src/SOS/Strike/platform/hostimpl.cpp:
```diff
@@ -0,0 +1,105 @@
+#ifdef FEATURE_PAL
+    return HostType::Lldb;
+#else
+    return HostType::DbgEng;
```
Yes, dotnet-dump (and eventually the VS plugin someday) will provide an IHost.

mikem8361 replied:
hostimpl.cpp will only be used in lldb/windbg when the managed host can't be started in the future. But in this PR before the upcoming part 3, it is a little broken since it is always used when dotnet-dump provides it. It should be ok for now.

tommcdon commented:
FYI @willpe
